onnxruntime/core/language_interop_ops/torch/custom_function_register.h

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include <mutex>
+#include <Python.h>
+#include "core/platform/env.h"
+
+namespace onnxruntime {
+namespace language_interop_ops {
+namespace torch {
+
+class OrtTorchFunctionPool final {
+ public:
+  static OrtTorchFunctionPool& GetInstance() {
+    static OrtTorchFunctionPool instance_;
+    return instance_;
+  }
+  void RegisterObject(PyObject* obj);
+  void RegisterForwardCore(const std::string& key, PyObject* obj);
+  void RegisterBackwardCore(const std::string& key, PyObject* obj);
+  void RegisterForwardRunner(PyObject* obj);
+  void RegisterBackwardRunner(PyObject* obj);
+  PyObject* GetForwardRunner();
+  PyObject* GetBackwardRunner();
+  PyObject* GetForwardCore(const std::string& key);
+  PyObject* GetBackwardCore(const std::string& key);
+
+  int64_t RegisterContext(PyObject* auto_grad_context);
+
+  PyObject* GetContext(int64_t context_index);
+  void UnRegisterContext(int64_t context_index);
+
+ private:
+  OrtTorchFunctionPool() = default;
+  ~OrtTorchFunctionPool() = default;
+  OrtTorchFunctionPool(const OrtTorchFunctionPool&) = delete;
+
+  OrtTorchFunctionPool& operator=(const OrtTorchFunctionPool&) = delete;
+
+  std::vector<PyObject*> obj_pool;
+  PyObject* forward_runner;
+  PyObject* backward_runner;
+  std::unordered_map<std::string, PyObject*> forward_core_pool;
```

Comment:
a comment for difference between core_pool and pool ? 
and what are they used for?

Reply:
Done.

Reply:
the pool is not needed any more. just keep core_pool for now. Maybe we don't need the "core"? @wschin

Reply:
`obj_pool` will finally be removed after we have RAII-style refcnt control. It's currently used for extending life-time of some Python objects.

Reply:
revisit all those refcnt incresement. added comments. removed useless fields for example obj_pool.